File: tests/hardware/unit/test_hazard_detection_unit.py
import cocotb
from cocotb.triggers import Timer
import sys
import os

# Import infrastructure
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from infrastructure import run_test_simple

@cocotb.test()
async def hazard_detection_unit_test(dut):
    """Test Hazard Detection Unit."""

    # Helper to check output
    async def check(rs1_id, rs2_id, rd_ex, mem_read_ex, expected_stall, name):
        dut.rs1_index_decode.value = rs1_id
        dut.rs2_index_decode.value = rs2_id
        dut.rd_index_execute.value = rd_ex
        dut.memory_read_enable_execute.value = mem_read_ex
        
        await Timer(1, units="ns")
        
        got = int(dut.stall_pipeline.value)
        assert got == expected_stall, f"{name} Failed: Expected={expected_stall}, Got={got}"

    # 1. No Hazard (No Load in EX)
    await check(1, 2, 3, 0, 0, "No Hazard (No Load)")

    # 2. No Hazard (Load in EX, but no dependency)
    await check(1, 2, 3, 1, 0, "No Hazard (Load, No Dep)")

    # 3. Load-Use Hazard on RS1
    # EX is Load to R1. ID uses R1.
    await check(1, 2, 1, 1, 1, "Hazard RS1")

    # 4. Load-Use Hazard on RS2
    # EX is Load to R2. ID uses R2.
    await check(1, 2, 2, 1, 1, "Hazard RS2")

    # 5. x0 Check (Never stall for x0)
    # EX is Load to x0 (unlikely but possible in logic). ID uses x0.
    await check(0, 2, 0, 1, 0, "x0 Hazard Check")

    dut._log.info("Hazard Detection Unit Test Passed!")

def test_hazard_detection_unit():
    run_test_simple(
        module_name="test_hazard_detection_unit",
        toplevel="hazard_detection_unit",
        rtl_files=["core/backend/hazard_detection_unit.v"],
        file_path=__file__
    )

File: test/unit_test/test_hazard_detection_unit.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
#include "tb_base.h"
#include "Vhazard_detection_unit.h"

/**
 * Hazard Detection Unit Testbench
 * Detects load-use hazards that require pipeline stall
 */
class HazardDetectionTestbench : public TestbenchBase<Vhazard_detection_unit> {
public:
    HazardDetectionTestbench() : TestbenchBase<Vhazard_detection_unit>(false) {
    }
    
    void check(uint8_t rs1_id, uint8_t rs2_id, uint8_t rd_ex, uint8_t mem_read_ex, uint8_t expected_stall, const char* name) {
        dut->rs1_index_decode = rs1_id;
        dut->rs2_index_decode = rs2_id;
        dut->rd_index_execute = rd_ex;
        dut->memory_read_enable_execute = mem_read_ex;
        eval();
        
        INFO(name);
        CHECK(dut->stall_pipeline == expected_stall);
    }
    
    void test_no_hazard() {
        check(1, 2, 3, 0, 0, "No Hazard (No Load)");
        check(1, 2, 3, 1, 0, "No Hazard (Load, No Dep)");
    }
    
    void test_load_use_hazard() {
        check(1, 2, 1, 1, 1, "Hazard on RS1");
        check(1, 2, 2, 1, 1, "Hazard on RS2");
        check(1, 1, 1, 1, 1, "Hazard on both RS1==RS2");
    }
    
    void test_x0_no_stall() {
        check(0, 2, 0, 1, 0, "x0 Hazard Check RS1");
        check(1, 0, 0, 1, 0, "x0 Hazard Check RS2");
        check(0, 0, 0, 1, 0, "x0 Hazard Check Both");
    }
};

TEST_CASE("Hazard Detection Unit") {
HazardDetectionTestbench tb;
        
        tb.test_no_hazard();
        tb.test_load_use_hazard();
        tb.test_x0_no_stall();
}
